File: npmjs/pray-cli/src/resolve/upstream.ts
import { PrayError } from "../errors.js";
import type { LockedUpstream, Lockfile } from "../lockfile/types.js";
import type { ManifestPackage, ManifestSource } from "../manifest/types.js";
import type { PackageSpec } from "../package-spec/types.js";
import type { ResolveOptions } from "./context.js";
import { impliedSourceName } from "./package-root.js";
import type { ResolvedPackage } from "./types.js";

type ResolveNamed = (declaration: ManifestPackage) => Promise<ResolvedPackage>;

interface UpstreamRefreshCandidate {
  declaration: Pick<ManifestPackage, "name" | "path">;
  spec: Pick<PackageSpec, "upstream">;
}

export function assertPathUpstreamRefreshSupported(
  packages: readonly UpstreamRefreshCandidate[],
  selectedPackage?: string,
): void {
  const unsupported = packages.find(
    (packageEntry) =>
      packageEntry.declaration.path &&
      packageEntry.spec.upstream &&
      (!selectedPackage || packageEntry.declaration.name === selectedPackage),
  );
  if (unsupported) {
    throw PrayError.unsupported(
      `this installation cannot refresh upstream package ${unsupported.declaration.name}; install pray with Cargo and retry`,
    );
  }
}

export async function lockPathUpstream(
  declaration: ManifestPackage,
  spec: PackageSpec,
  sources: Map<string, ManifestSource>,
  lockfile: Lockfile | undefined,
  options: ResolveOptions,
  resolveNamed: ResolveNamed,
): Promise<LockedUpstream | undefined> {
  const upstream = spec.upstream;
  if (!upstream) {
    return undefined;
  }
  if (spec.name === upstream.name) {
    throw PrayError.resolution(
      `package ${spec.name} cannot use itself as upstream`,
    );
  }
  if (!declaration.path) {
    return undefined;
  }
  const refresh =
    options.ignoreLockedVersions ||
    options.unlockedPackages.has(declaration.name);
  const locked = lockfile?.package.find(
    (entry) => entry.name === declaration.name,
  )?.upstream;
  if (!refresh && locked && locked.name !== upstream.name) {
    throw PrayError.integrity(
      `locked upstream name mismatch for ${declaration.name}: expected ${locked.name}, found ${upstream.name}`,
    );
  }
  const name = !refresh && locked ? locked.name : upstream.name;
  const constraint =
    refresh || !locked ? upstream.constraint : `= ${locked.version}`;
  const source =
    !refresh && locked
      ? locked.source
      : impliedSourceName(
          {
            name,
            constraint,
            exports: [],
            targets: [],
            features: [],
            groups: [],
            optional: false,
          },
          sources,
        );
  const resolved = await resolveNamed({
    name,
    constraint,
    ...(source ? { source } : {}),
    exports: [],
    targets: [],
    features: [],
    groups: [],
    optional: false,
    roles: [],
    bound: false,
  });
  const resolvedUpstream: LockedUpstream = {
    name,
    version: resolved.spec.version,
    ...(source ? { source } : {}),
    tree_hash: resolved.treeHash,
    artifact_hash: resolved.artifactHash,
  };
  if (!refresh && locked) {
    ensureLockedUpstreamMatches(locked, resolvedUpstream);
  }
  return resolvedUpstream;
}

export function ensureLockedUpstreamMatches(
  locked: LockedUpstream,
  resolved: LockedUpstream,
): void {
  if (locked.name !== resolved.name || locked.version !== resolved.version) {
    throw PrayError.integrity("locked upstream identity mismatch");
  }
  if (locked.source !== resolved.source) {
    throw PrayError.integrity("locked upstream source mismatch");
  }
  if (locked.tree_hash !== resolved.tree_hash) {
    throw PrayError.integrity("locked upstream tree hash mismatch");
  }
  if (locked.artifact_hash !== resolved.artifact_hash) {
    throw PrayError.integrity("locked upstream artifact hash mismatch");
  }
}

File: rubygems/pray-cli/lib/pray/upstream.rb
# frozen_string_literal: true

module Pray
  module Upstream
    module_function

    def ensure_update_supported!(packages, selected = nil)
      unsupported = packages.find do |package|
        package.declaration.path && package.spec.upstream &&
          (selected.nil? || package.declaration.name == selected)
      end
      return unless unsupported

      raise Error.unsupported(
        "this installation cannot refresh upstream package #{unsupported.declaration.name}; " \
        "install pray with Cargo and retry"
      )
    end

    def to_lock_hash(entry)
      hash = {
        "name" => entry.name,
        "version" => entry.version,
        "tree_hash" => entry.tree_hash,
        "artifact_hash" => entry.artifact_hash
      }
      hash["source"] = entry.source unless entry.source.nil?
      hash
    end

    def from_lock_hash(entry)
      return nil if entry.nil?

      LockedUpstream.new(
        name: entry["name"],
        version: entry["version"],
        source: entry["source"],
        tree_hash: entry["tree_hash"],
        artifact_hash: entry["artifact_hash"]
      )
    end

    def lock_path(project_root, sources, git_sources, user_config, declaration, spec, lockfile, options)
      return nil unless spec.upstream
      if spec.name == spec.upstream.name
        raise Error.resolution("package #{spec.name} cannot use itself as upstream")
      end
      return nil unless declaration.path

      refresh = options.ignore_locked_versions || options.unlocked_packages.include?(declaration.name)
      locked = lockfile&.package&.find { |entry| entry.name == declaration.name }&.upstream
      name, constraint, source = resolution_input(declaration, spec, sources, locked, refresh)
      resolved = Resolve.resolve_package(
        project_root,
        sources,
        git_sources,
        user_config,
        ManifestPackage.new(
          name: name,
          constraint: constraint,
          source: source
        ),
        lockfile,
        options: options
      )
      resolved_upstream = LockedUpstream.new(
        name: name,
        version: resolved.spec.version,
        source: source,
        tree_hash: resolved.tree_hash,
        artifact_hash: resolved.artifact_hash
      )
      ensure_locked_match!(locked, resolved_upstream) if !refresh && locked
      resolved_upstream
    end

    def resolution_input(declaration, spec, sources, locked, refresh)
      if locked && !refresh
        if locked.name != spec.upstream.name
          raise Error.integrity(
            "locked upstream name mismatch for #{declaration.name}: " \
            "expected #{locked.name}, found #{spec.upstream.name}"
          )
        end
        return [locked.name, "= #{locked.version}", locked.source]
      end

      name = spec.upstream.name
      source = ResolveSource.implied_source_name(ManifestPackage.new(name: name), sources)
      [name, spec.upstream.constraint, source]
    end

    def ensure_locked_match!(locked, resolved)
      if locked.name != resolved.name || locked.version != resolved.version
        raise Error.integrity("locked upstream identity mismatch")
      end
      raise Error.integrity("locked upstream source mismatch") if locked.source != resolved.source
      raise Error.integrity("locked upstream tree hash mismatch") if locked.tree_hash != resolved.tree_hash
      if locked.artifact_hash != resolved.artifact_hash
        raise Error.integrity("locked upstream artifact hash mismatch")
      end
    end
  end
end
